src/ui/ui_i18n/README.md

```diff
@@ -0,0 +1,325 @@
+  `uiExports.translations`), then that locale will be resolved.
+- If locale data is missing for a leaf locale like `fr-FR`, but there is data
+for one of its ancestors, `fr` in this case, then its ancestor will be used.
+- If `accept-language` header is not presented or previous steps didn't resolve
+the locale, the locale will be resolved to locale defined in `i18n.defaultLocale`
+option at `config/kibana.yml` file.
+- If `i18n.defaultLocale` is not specified or if there is no data for this locale,
+`en` locale will be used.
+
+## I18n engine
+
+I18n engine is the platform agnostic abstractions which helps to supply a locale
+data to UI frameworks and provides methods for the direct translation.
+
+Here is the public API exposed by this engine:
+
+- `getMessages()` - returns messages for the current language
+- `setLocale(locale: string)` - tells the engine which language to use by given
+language key
+- `getLocale()` - returns the current locale
+- `setDefaultLocale(locale: string)` - tells the library which language to fallback
+when missing translations
+- `getDefaultLocale()` - returns the default locale
+- `defineFormats(formats: object)` - supply a set of options to the underlying formatter.
+For the detailed explanation, see the section below
+- `translate(id: string, [{values: object, defaultMessage: string}])` – translate message by id
+
+#### I18n engine internals
+
+The engine uses the ICU Message syntax and works for all CLDR languages which
+have pluralization rules defined. It's built around `intl-messageformat` package
+which exposes `IntlMessageFormat` class. Messages are provided into the constructor
+as a string message, or a pre-parsed AST object.
+
+```js
+import IntlMessageFormat from 'intl-messageformat';
+
+const msg = new IntlMessageFormat(message, locales, [formats]);
+```
+
+The string `message` is parsed, then stored internally in a
+compiled form that is optimized for the `format()` method to
+produce the formatted string for displaying to the user.
+
+```js
+const output = msg.format(values);
+```
+
+`formats` parameter in `IntlMessageFormat` constructor allows to format numbers
+and dates/times in messages using `Intl.NumberFormat` and `Intl.DateTimeFormat`,
+respectively.
+
+```js
+const msg = new IntlMessageFormat('The price is: {price, number, USD}', 'en-US', {
+  number: {
+    USD: {
+      style   : 'currency',
+      currency: 'USD',
+    },
+  },
+});
+
+const output = msg.format({ price: 100 });
+
+console.log(output); // => "The price is: $100.00"
+```
+
+In this example, we're defining a USD number format style which is passed to
+the underlying `Intl.NumberFormat` instance as its options.
+[Here](https://github.com/yahoo/intl-messageformat/blob/master/src/core.js#L62)
+you can find default format options used as the prototype of the formats
+provided to the constructor.
+
+Creating instances of `IntlMessageFormat` is an expensive operation.
```

Comment:
**grammar:** Creating instances of `IntlMessageFormat` is expensive. (Mainly because each instance creation is expensive.)